[GEN][ZH] Center Game App Window on startup and resolution change

helmutbuhler opened the pull request:
This PR centers the window after loading on the screen (not including taskbar).
It also adds code to pump window messages during startup to avoid "(Not responding)" in the title. (That was necessary because Windows otherwise ignores the Move Call in debug builds, and maybe also on slower machines)

Also note that I needed to #define WINVER 0x0500, because the Monitor API didn't exist in Win 95 yet. This causes the compiler to emit a message when compiling dx8wrapper.cpp, but generates no error or warning. The game will also no longer run on Win95, but I think we can live with that.

## Change list

- fix: In windowed mode, the game now starts centered on the screen.

## TODO

- [x] Replicate in Generals

tintinhamans commented:
While this is neat, how do we want to handle cases where the user does not want this functionality.

tomsons26 commented:
I'm gonna also ruin this and say, you will likely only center the initial window, the game makes two, one to show splash other for game, we tried this in Thyme but result was only splash window got centered

helmutbuhler commented:
> While this is neat, how do we want to handle cases where the user does not want this functionality.

The current behavior is that the window starts in the bottom right corner. The bottom right part for the window is usually not visible. Tell me who wants this behavior?
Also, Gentool changes the behavior to what this PR implements, did anyone complain about that?



> I'm gonna also ruin this and say, you will likely only center the initial window, the game makes two, one to show splash other for game, we tried this in Thyme but if the splash window doesn't match the res of the game, result was only splash window got centered

Are you saying that this PR is wrong because you didn't get this feature to work in Thyme? If you look at the code or try out the build, you'll see that it works.

xezon commented:
> Also, Gentool changes the behavior to what this PR implements, did anyone complain about that?

No complaints over 15 years. Many users use the "Full" window preset in GenTool together with windowed mode.

tintinhamans commented:
Something doesn't seem to be quite right.. there seems to be a gap present below the window despite the correct height of 1440 being set in `options.ini`.

![NVIDIA_Overlay_4NI2GDvNVx(1)](https://github.com/user-attachments/assets/8d926620-d7b8-4e9f-9be1-e3fcdcc085a0)

tintinhamans commented:
> No complaints over 15 years. Many users use the "Full" window preset in GenTool together with windowed mode.

I can't use Full, it pulls the window to the left of my monitor 😢 :

![image](https://github.com/user-attachments/assets/70082d82-6ea6-479b-8e8f-b8ba55dc1d7b)

tomsons26 commented:
> > While this is neat, how do we want to handle cases where the user does not want this functionality.
> 
> The current behavior is that the window starts in the bottom right corner. The bottom right part for the window is usually not visible. Tell me who wants this behavior? Also, Gentool changes the behavior to what this PR implements, did anyone complain about that?
> 
> > I'm gonna also ruin this and say, you will likely only center the initial window, the game makes two, one to show splash other for game, we tried this in Thyme but if the splash window doesn't match the res of the game, result was only splash window got centered
> 
> Are you saying that this PR is wrong because you didn't get this feature to work in Thyme? If you look at the code or try out the build, you'll see that it works.

I'm saying there may be unexpected behavior that may not always manifest due to the game making and destroying multiple windows during its startup

willherzog commented:
The new official patch (1.05) adds window centering. I have no idea whether it's done in the same way as this PR, but certainly for me it's much appreciated after having to deal for so long with the game window starting out not only off-centered but also partially off-screen.

Technically what appeared to happen originally is that the splash screen was centered but then the actual game window used the exact same top+left coordinates as the splash screen - so any game resolution besides 800x600 (which would be the same as the splash screen) necessarily resulted in it being off-centered.

helmutbuhler commented:
Never mind, I found the issue. I center the window in the workarea (area not occupied by taskbar and similar), but if the resolution is bigger than the workarea, it's better to center in the whole monitorarea. I just pushed a commit to fix this.
While I was at it, I also changed the codepath that repositions the window after changing the resolution to do the same. I also added some more comments.

@tintinhamans Can you check if the behavior is now fixed?

OmniBlade reviewed (commented):
If the previous issues reported are confirmed fixed I'm happy to approve this?

helmutbuhler commented:
Yes, please.

aliendroid1 commented:
I think the main reason for the issue is that the position is set based on the launcher image resolution but then not updated when resolution is updated

helmutbuhler commented:
> I think the main reason for the issue is that the position is set based on the launcher image resolution but then not updated when resolution is updated

Yes, and this PR fixes it

xezon reviewed (commented):
I tested this change and the behaviour this adds is very good. It automatically centers or fullscreens the game window depending on the resolution.

It also updates the position when changing the resolution from the Options menu. The window reposition does look a bit buggy then, but that is unrelated to this change. For reference, with GenTool the window reposition after resolution change looks a bit better.

I do not understand what bug the `serviceWindowsOS` call fixes. I cannot reproduce it.

This change needs to be replicated in Generals.

helmutbuhler commented on GeneralsMD/Code/Libraries/Source/WWVegas/WW3D2/dx8wrapper.cpp:
```diff
@@ -44,7 +44,8 @@
 #define WW3D_DEVTYPE D3DDEVTYPE_HAL
-
+#undef WINVER
+#define WINVER 0x0500
```
This is only for VC6

xezon replied:
Then put it in VC6 ifdef.

helmutbuhler replied:
fixed

helmutbuhler commented on GeneralsMD/Code/Libraries/Source/WWVegas/WW3D2/dx8wrapper.cpp:
```diff
@@ -885,6 +906,60 @@ void DX8Wrapper::Get_Format_Name(unsigned int format, StringClass *tex_format)
+			MONITORINFO mi = {sizeof(MONITORINFO)};
+			GetMonitorInfo(MonitorFromWindow(_Hwnd, MONITOR_DEFAULTTOPRIMARY), &mi);
+			int left = (mi.rcWork.left + mi.rcWork.right - width) / 2;
+			int top  = (mi.rcWork.top + mi.rcWork.bottom - height) / 2;
```
It should work, I tested it a bit on a 2-monitor setup.

helmutbuhler commented on GeneralsMD/Code/GameEngine/Source/Common/GameEngine.cpp:
```diff
@@ -371,6 +371,13 @@ void GameEngine::init( int argc, char *argv[] )
+		// Pump messages during startup to avoid "(Not responding)" in the window title
+		// on slower computers and in debug build.
+		// This also ensures that the window is correctly positioned, because Windows
+		// apparently ignores the SetWindowPos call when the window is not responding.
```
Well, Windows behaves as if it ignores the call. Does it actually ignore it, or does the GhostWindow implementation have bugs that ignore that call, or is it some other bug? I don't know, I don't have the code. All I know is that Windows appears to ignore the call, so I think the comment is accurate.

xezon replied:
Ok if we do not know what Windows really does, then how about we just describe the observed issue? This way there will be no speculation.

Example: "Windows behaves as if the call to SetWindowPos never occurred".

aliendroid1 replied:
> Ok if we do not know what Windows really does, then how about we just describe the observed issue? This way there will be no speculation.
> 
> Example: "Windows behaves as if the call to SetWindowPos never occurred".

I think the mistake causing this issue is inside the d3d8 wrapper setRenderDevice function. Even when passed the argument to resize, inside it's using flags for no resize and no reposition

helmutbuhler replied:
@xezon In my opinion the comment is fine as it is and means the same as your suggestion. If you feel the comment needs to be changed, can you do it?

@aliendroid1 No, the flags are correct.

helmutbuhler commented:
Thanks for the review! I believe everything is resolved now. I can also sync with Generals once approved.

xezon commented on GeneralsMD/Code/GameEngine/Source/Common/GameEngine.cpp:
```diff
@@ -371,6 +371,13 @@ void GameEngine::init( int argc, char *argv[] )
+		// on slower computers and in debug build.
+		// This also ensures that the window is correctly positioned, because Windows
+		// apparently ignores the SetWindowPos call when the window is not responding.
+		serviceWindowsOS();
```
I cannot confirm that this avoids "(Not Responding)" in the window title. I tested this with

```cpp
serviceWindowsOS();
Sleep(10000);
serviceWindowsOS();
```

But it does fix the window positioning as you say.

xezon commented:
Comments simplified. Two new debug logs added. Replicated in Generals. Check if good.

helmutbuhler commented:
Thanks! Looks all good!

I left a comment above about that "Not responding" things, in case you are curious.

Slightly related to this PR is also this: https://github.com/TheSuperHackers/GeneralsGameCode/pull/595 Maybe you can approve that, too? (I already reviewed it)

xezon commented:
Appended a few more minor word tweaks.
